JS Public API Update

rtibbles opened the pull request:
**N.B. I strongly advise you to only review this commit by commit (skipping the very large commits that move all files and update all imports) - opening the Files changed tab for all the changes here will probably crash your tab, if not your browser.**

## Summary
For the new API this implements, see #5488 

* Creates migration script that reads file imports in source code and maps them to the updated import paths - available as kolibri-tools command `migrate`
  * Shows a warning to a user if a mapped import maps to null
  * Shows a warning for jest.mocks where the module mapping is more complex, as the mocks will need to be resolved manually
* Migrates every APIResource into its own module, all except TaskResource, is in kolibri-common
* Migrate the new independent packages into their own packages in the kolibri packages directory
* Creates a kolibri package
  * migrates all public modules there
  * sets exports explicitly in the package.json to prevent access to internal only modules
  * adds an 'exposes' field that details the external modules the core JS API also supports

* Migrates all other common modules to kolibri-common
* Removes all core aliases
  * Removes the use of the custom import resolver in our linting configuration, but retains it for use in a couple of other cases for now
* Updates core externals to map to a spec object defined in the new kolibri package
* Updates the spec export tools to introspect the kolibri package package.json to generate the apiSpec file
  * Use require statements exclusively to ensure we are always exposing the full module object
  * Add a special module resolution for the vue composition API plugin that managed to break this due to its janky internal module configuration

## References
Fixes #5488

## Reviewer guidance
The most changes here are in two commits, one that moves all the files to their new locations, and the other that updates all the import references. I would strongly recommend not viewing these commits in a browser interface.

Aside from this, absolutely nothing should have changed from a functional point of view, and all tests should be passing. If anything is awry, then that will need fixing.

There may be minor deviations from the spec, as it became a little tiresome to keep them in sync, please flag if anything looks suss, and I can either explain or rectify. Also happy to hear comments where the spec doesn't make as much sense as it might!

Lastly, this is not a finalization of the api spec - I suspect that we can finish removing the core vuex, and hence remove it from the core API completely, and if we want to break up, say, commonCoreStrings into more explicit translator modules, that also seems like a good idea.

Two other open questions I had were about the `kolibri` package:

1. For all the other new packages I created, I put everything inside a `src` directory inside the package, I am not sure why I didn't do the same for the `kolibri` package, but it would be straight forward enough to do and then update the `exports` field accordingly.
2. Is it obvious enough which things are private and which are public in this package? i.e. by using the exports field, or would it be better to try to make it more obvious by either prefixing private modules with `_`, or putting the api files into a special folder, and the private ones elsewhere?

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

rtibbles committed:
Add new base kolibri packages.

rtibbles committed:
Add tooling for migrating core API imports.

rtibbles committed:
Migrate code to kolibri and kolibri-common

rtibbles committed:
Remove custom aliases from import resolution.

rtibbles committed:
Migrate all core API import references.
Empty core API.

rtibbles committed:
Explicitly define kolibri module API.
Update core externals generation.
Add automatically generated API spec.
Update core specific module replacements.

rtibbles commented:
Switching to draft - something has gone awry in the build that I hadn't seen previously.

rtibbles committed:
Clean up previously unflagged import linting warnings.
Remove extraneous getters now covered by the useUser composable.

rtibbles commented:
Issues resolved.

marcellamaki commented on kolibri/core/package.json:
```diff
@@ -4,37 +4,6 @@
-    "vuex": "^3.6.2",
-    "vuex-router-sync": "^5.0.0",
-    "xstate": "^4.38.3"
+    "core-js": "^3.38"
```
So what's happening in this commit generally "Add new base kolibri packages" is that for everything that got moved out of kolibri core, where the dependencies were required in a moved file, the package.json for that is in the plugin where the components that require those dependencies are located? Or the packages, or kolibri-common.

rtibbles replied:
Yes - now this package.json only includes things that are only imported in the default frontend bundle (which is very little). Most of the imports are used either in the `kolibri` package or the `kolibri-common` package, and so are now specified there.

marcellamaki commented on packages/kolibri-tools/lib/webpack.config.plugin.js:
```diff
@@ -172,11 +167,27 @@ module.exports = (
+      // conditionally imports the prod or dev version of the package based on the NODE_ENV, but
+      // only exposes the common JS builds, and reassigns them to the module.exports, thereby
+      // preventing webpack from importing it consistently in both how we use it internally
+      // in the core bundle, and also then access it via externals in the plugin bundles.
```
I am following this comment 80% and I think I understand in broad strokes what is happening, but if I were to really explain the _why_, I don't think I would get it. The two phrases that I keep getting tripped up on are 
> only exposes the common JS builds, and reassigns them to the module.exports

and
> also then access it via externals in the plugin bundles

rtibbles replied:
Basically, internally, the composition api does something really janky for its exports, which breaks how we expose the 'core' modules on the default frontend bundle for the plugins to reuse without having to bundle them. This is a workaround for that.

If it makes you feel any better, once we've upgraded to Vue 2.7, we can get rid of this entirely, because we won't need the plugin any more!

marcellamaki commented on packages/kolibri/apiSpec.js:
```diff
@@ -1,5 +1,97 @@
  */
 
-export default {};
+const kolibri_apiResource = require('kolibri/apiResource');
```
I don't know why we are using the `_` here. I don't have any strong opinions about why we should or shouldn't, but just wondering what the thinking is. we have updated from the prevoius `.` (i.e. the update from kolibri.coreVue.AppBar or whatnot). But I don't know if it's about having a clear delineation in time that this has happened? or if it's just about a preference for snake case? or if it's somehow similar to the leading `__` for private functions? Or other connection to this question/comment in the reviewer guidance
>Is it obvious enough which things are private and which are public in this package

I just feel like there's something I missing here about this change, when reading it

rtibbles replied:
Hrm, I'll change how I am generating the apiSpec, I can just make the requires inline without needing the intermediary variables. You're getting tripped up on the variable name containing a `_` it seems, but that is not exposed anywhere outside of this file, it is entirely internal.

marcellamaki reviewed (commented):
I've only added a couple of comments so far and am still digesting. I haven't decided yet about my opinion on 
> i.e. by using the exports field, or would it be better to try to make it more obvious by either prefixing private modules with _, or putting the api files into a special folder, and the private ones elsewhere?

and will come back to that later with some thoughts. 

Not blocking for this PR but I think it would be very worthwhile to update [the dev docs](https://kolibri-dev.readthedocs.io/en/develop/frontend_architecture/core.html#shared-core-functionality) both for accuracy and also to have some guidance for common vs. core, and in what circumstances we will make updates to the core API spec (at least as we move towards/before we get to v1.x)

rtibbles committed:
Update developer documentation for core API changes.

marcellamaki commented:
Follow up on the specific questions asked: 

> For all the other new packages I created, I put everything inside a src directory inside the package, I am not sure why I didn't do the same for the kolibri package, but it would be straight forward enough to do and then update the exports field accordingly.

I don't have a strong opinion on this, but I think for the sake of consistency, having it also in a /src package makes sense to me.

> Is it obvious enough which things are private and which are public in this package? i.e. by using the exports field, or would it be better to try to make it more obvious by either prefixing private modules with _, or putting the api files into a special folder, and the private ones elsewhere?

I would say after reading through this a couple of times, it's not obvious. It's not hard to sort out once you've been looking around a little bit, but I think we could make it easier. Prefixing with _ seems like it would be easier to tell at a glance when looking at a specific file, but a separate folder would make more sense to me if I were coming in fresh and trying to understand the project structure overall. Which makes me think...both might be useful? Admittedly I know less about coding conventions, so was there a specific reason you suggested "either"? Or just that doing both feels unnecessary.

rtibbles commented:
> Or just that doing both feels unnecessary.

It felt a little unnecessary, but also both are completely doable, and would help signpost it more, if you happened to be in a deep folder structure.

The only real issue with separate `public` and `internal` folders is just that the public files will end up having to do a lot of long relative path imports, and code that is used together doesn't naturally live together, so if I was only going to do one of these things, I would prefer it to be prefixing with `_` so that we could keep the existing code layout.

rtibbles commented:
> I don't have a strong opinion on this, but I think for the sake of consistency, having it also in a /src package makes sense to me.

In that case, shall I move all the files in `kolibri-common` also into `/src` - we can do so without updating paths, by doing a wildcard export to expose the `src` folder as the top level.

bjester commented:
On the topic of public vs private, the `exports` field, and prefixing `_` to explicitly distinguish private code, could using the `_` convention allow us to write a linting rule that ensures the `exports` field is kept up-to-date?

marcellamaki commented:
>  the public files will end up having to do a lot of long relative path imports, and code that is used together doesn't naturally live together

Right, I hadn't really thought about that. I think prefixing is fine then. And I will reread the documentation updates, and see if there's anything else that I can think of that might be worth adding there to help.

bjester commented:
Re: public vs private and exports

It appears this feature is also possible, according to [node.js docs](https://nodejs.org/api/packages.html#package-entry-points), where you specify `null` to identify what's internal.
```json
{
  "name": "my-package",
  "exports": {
    ".": "./lib/index.js",
    "./feature/*.js": "./feature/*.js",
    "./feature/internal/*": null
  }
} 
```

bjester commented:
> For all the other new packages I created, I put everything inside a src directory inside the package, I am not sure why I didn't do the same for the kolibri package, but it would be straight forward enough to do and then update the exports field accordingly.

Consistency always feels good and helps to clearly distinguish from package tooling or other settings, from the actual source code, so I support this

rtibbles commented:
> On the topic of public vs private, the exports field, and prefixing _ to explicitly distinguish private code, could using the _ convention allow us to write a linting rule that ensures the exports field is kept up-to-date?

Yes, I think so - and we could even have the exports field auto generated similarly to the way that the apiSpec file is autogenerated from the exports and exposes fields.

rtibbles commented:
> It appears this feature is also possible

Interesting, so technically you could do a wild card instead of `feature` and have it so any folder labelled with `internal` is not exportable.

bjester commented on kolibri/plugins/device/assets/src/views/__test__/RearrangeChannelsPage.spec.js:
```diff
@@ -1,12 +1,12 @@
-import useUser, { useUserMock } from 'kolibri.coreVue.composables.useUser';
-import useSnackbar, { useSnackbarMock } from 'kolibri.coreVue.composables.useSnackbar';
+import useUser, { useUserMock } from 'kolibri/composables/useUser'; // eslint-disable-line
+import useSnackbar, { useSnackbarMock } from 'kolibri/composables/useSnackbar'; // eslint-disable-line
```
Why does eslint not like these lines?

rtibbles replied:
Because the named exports only exist inside the mock, not inside the composable module itself, but because it is being mocked, they are available as an export.

bjester commented on kolibri/plugins/learn/package.json:
```diff
@@ -3,6 +3,10 @@
   "private": true,
   "version": "0.0.1",
+  "dependencies": {
+    "deep-object-diff": "^1.1.9",
```
Not related to your work here, but I'm curious if `lodash` can accomplish what this dependency does

rtibbles replied:
Yes, looking at the usage in place, it is only being used to check deep equality, so lodash's isEqual would do the job here instead.

bjester reviewed (approved):
I spot checked a bunch of files but didn't find any concerns other than those I've already brought up before this. :+1:

rtibbles committed:
Move all internal modules in the kolibri package into folders called 'internal'.
Update the kolibri package build tools to leverage this fact to automatically generate the exports field.

rtibbles commented:
Added follow up issue for resolving the `src` folder work here: https://github.com/learningequality/kolibri/issues/12780

rtibbles committed:
Revert stylelint-config-sass-guidelines upgrade as it causes ESM errors.

indirectlylit commented:
💥  🎉
